File: system-test/dotnet/Driver.Adapter/Shop/Api/Client/ShopApiClient.cs
using Driver.Adapter.Shared.Client.Http;

using Driver.Adapter.Shop.Api.Client.Dtos.Errors;

using Driver.Adapter.Shop.Api.Client.Controllers;



namespace Driver.Adapter.Shop.Api.Client;



public class ShopApiClient : IDisposable

{

    private readonly JsonHttpClient<ProblemDetailResponse> _httpClient;

    private readonly HealthController _healthController;

    private readonly OrderController _orderController;

    private bool _disposed;



    public ShopApiClient(string baseUrl)

    {

        _httpClient = new JsonHttpClient<ProblemDetailResponse>(baseUrl);

        _healthController = new HealthController(_httpClient);

        _orderController = new OrderController(_httpClient);

    }



    public void Dispose()

    {

        Dispose(true);

        GC.SuppressFinalize(this);

    }



    protected virtual void Dispose(bool disposing)

    {

        if (_disposed) return;

        if (disposing)

            _httpClient?.Dispose();

        _disposed = true;

    }



    public HealthController Health() => _healthController;



    public OrderController Orders() => _orderController;





}

File: system-test/java/src/main/java/com/optivem/shop/dsl/driver/adapter/shop/api/client/ShopApiClient.java
package com.optivem.shop.dsl.driver.adapter.shop.api.client;

import com.optivem.shop.dsl.driver.adapter.shop.api.client.controllers.HealthController;
import com.optivem.shop.dsl.driver.adapter.shop.api.client.controllers.OrderController;
import com.optivem.shop.dsl.driver.adapter.shop.api.client.dtos.errors.ProblemDetailResponse;
import com.optivem.shop.dsl.driver.adapter.shared.client.http.JsonHttpClient;
import com.optivem.shop.dsl.common.Closer;

public class ShopApiClient implements AutoCloseable {
    private final JsonHttpClient<ProblemDetailResponse> httpClient;
    private final HealthController healthController;
    private final OrderController orderController;

    public ShopApiClient(String baseUrl) {
        this.httpClient = new JsonHttpClient<>(baseUrl, ProblemDetailResponse.class);
        this.healthController = new HealthController(httpClient);
        this.orderController = new OrderController(httpClient);
    }

    public HealthController health() {
        return healthController;
    }

    public OrderController orders() {
        return orderController;
    }

    @Override
    public void close() {
        Closer.close(httpClient);
    }
}



